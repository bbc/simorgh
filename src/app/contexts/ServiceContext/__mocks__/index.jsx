import React from 'react';
import { node, string } from 'prop-types';
import services from '#server/utilities/serviceConfigs';
import { getVariant } from '#lib/utilities/variantHandler';
import getLangOverride from '#lib/utilities/langHandler';
/*
 * This file is mocked by default to avoid having to handle
 * async behavior in tests across the application when
 * using service contexts.
 */
export const ServiceContext = React.createContext({});
export const ServiceContextProvider = ({
  children,
  pageLang,
  service,
  variant,
}) => {
  const dataKey =
    getLangOverride({ service, pageLang }) || getVariant({ service, variant });
  return (
    <ServiceContext.Provider value={services[service][dataKey]}>
      {children}
    </ServiceContext.Provider>
  );
};
ServiceContextProvider.propTypes = {
  children: node.isRequired,
  service: string.isRequired,
  variant: string,
  pageLang: string,
};
ServiceContextProvider.defaultProps = {
  pageLang: null,
  variant: 'default',
};
