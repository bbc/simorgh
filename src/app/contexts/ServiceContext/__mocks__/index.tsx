import React, { PropsWithChildren } from 'react';
import { node, string } from 'prop-types';
import getLangOverride from '../../../lib/utilities/langHandler';
import { getVariant } from '../../../lib/utilities/variantHandler';
import services from '../../../../server/utilities/serviceConfigs';
import { Services, Variants } from '../../../models/types/global';

interface Props {
  service: Services;
  variant?: Variants | null;
  pageLang?: string;
}

/*
 * This file is mocked by default to avoid having to handle
 * async behavior in tests across the application when
 * using service contexts.
 */
export const ServiceContext = React.createContext({});
export const ServiceContextProvider = ({
  children,
  pageLang,
  service,
  variant,
}: PropsWithChildren<Props>) => {
  const dataKey: Variants =
    getLangOverride({ service, pageLang }) || getVariant({ service, variant });
  return (
    <ServiceContext.Provider value={services[service][dataKey]}>
      {children}
    </ServiceContext.Provider>
  );
};
ServiceContextProvider.propTypes = {
  children: node.isRequired,
  service: string.isRequired,
  variant: string,
  pageLang: string,
};
ServiceContextProvider.defaultProps = {
  pageLang: null,
  variant: 'default',
};
