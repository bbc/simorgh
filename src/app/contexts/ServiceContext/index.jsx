import { createContext } from 'react';
import { node, string } from 'prop-types';
import services from '../../lib/config/services/loadableConfig';
import { getVariant } from '../../lib/utilities/variantHandler';
import getLangOverride from '../../lib/utilities/langHandler';
import variantPropType from '../../models/propTypes/variants';

export const ServiceContext = createContext({});

export const ServiceContextProvider = ({
  children,
  service,
  variant,
  pageLang,
}) => {
  const LoadableContextProvider = services[service];

  if (!LoadableContextProvider) {
    return null;
  }

  return (
    <LoadableContextProvider
      Context={ServiceContext}
      dataKey={
        getLangOverride({ service, pageLang }) ||
        getVariant({ service, variant })
      }
    >
      {children}
    </LoadableContextProvider>
  );
};

ServiceContextProvider.propTypes = {
  children: node.isRequired,
  pageLang: string,
  service: string,
  variant: variantPropType,
};

ServiceContextProvider.defaultProps = {
  pageLang: null,
  service: 'default',
  variant: 'default',
};
