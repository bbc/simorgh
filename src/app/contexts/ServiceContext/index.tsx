import React, { createContext, PropsWithChildren } from 'react';
import services from '../../lib/config/services/loadableConfig';
import { getVariant } from '../../lib/utilities/variantHandler';
import getLangOverride from '../../lib/utilities/langHandler';
import { Services, Variants } from '../../models/types/global';
import { ServiceConfig } from '../../models/types/serviceConfig';

export const ServiceContext = createContext<ServiceConfig>({} as ServiceConfig);

interface Props {
  service: Services;
  variant?: Variants | null;
  pageLang?: string;
}

export const ServiceContextProvider = ({
  children,
  service,
  variant,
  pageLang,
}: PropsWithChildren<Props>) => {
  const LoadableContextProvider = services[service];

  if (!LoadableContextProvider) {
    return null;
  }

  return (
    <LoadableContextProvider
      Context={ServiceContext}
      dataKey={
        getLangOverride({ service, pageLang }) ||
        getVariant({ service, variant })
      }
    >
      {children}
    </LoadableContextProvider>
  );
};
