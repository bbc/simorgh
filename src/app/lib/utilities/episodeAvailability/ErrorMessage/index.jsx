import React, { useContext } from 'react';
import { oneOf } from 'prop-types';
import styled from '@emotion/styled';
import pathOr from 'ramda/src/pathOr';
import {
  GEL_SPACING,
  GEL_SPACING_DBL,
  GEL_SPACING_TRPL,
  GEL_SPACING_QUAD,
  GEL_SPACING_QUIN,
} from '#psammead/gel-foundations/src/spacings';
import {
  GEL_GROUP_1_SCREEN_WIDTH_MAX,
  GEL_GROUP_2_SCREEN_WIDTH_MAX,
  GEL_GROUP_3_SCREEN_WIDTH_MAX,
} from '#psammead/gel-foundations/src/breakpoints';

import { MediaMessage } from '#components/MediaPlayer';
import { EPISODE_STATUS } from '..';
import { ServiceContext } from '../../../../contexts/ServiceContext';

const AudioErrorWrapper = styled.div`
  position: relative;
  min-height: 165px;
  margin-bottom: ${GEL_SPACING_QUAD};
`;

const landscapeRatio = '56.25%'; // (9/16)*100 = 16:9
const VideoErrorWrapper = styled.div`
  margin: ${GEL_SPACING_QUIN} 0 ${GEL_SPACING_TRPL};
  padding-top: ${landscapeRatio};
  position: relative;
  overflow: hidden;

  @media (max-width: ${GEL_GROUP_3_SCREEN_WIDTH_MAX}) {
    margin-top: ${GEL_SPACING_DBL};
  }
  @media (max-width: ${GEL_GROUP_2_SCREEN_WIDTH_MAX}) {
    margin: ${GEL_SPACING_DBL} -${GEL_SPACING_DBL} 0;
  }
  @media (max-width: ${GEL_GROUP_1_SCREEN_WIDTH_MAX}) {
    margin: ${GEL_SPACING} -${GEL_SPACING} 0;
  }
`;

const getErrorMessage = (status, translations) => {
  if (status === EPISODE_STATUS.EPISODE_IS_EXPIRED) {
    return pathOr(
      'This content is no longer available',
      ['media', 'contentExpired'],
      translations,
    );
  }
  return pathOr(
    'This content is not yet available',
    ['media', 'contentNotYetAvailable'],
    translations,
  );
};

const ErrorMessage = ({ episodeAvailability, skin }) => {
  const { service, translations } = useContext(ServiceContext);
  const Wrapper = skin === 'audio' ? AudioErrorWrapper : VideoErrorWrapper;

  return (
    <Wrapper>
      <MediaMessage
        service={service}
        message={getErrorMessage(episodeAvailability, translations)}
      />
    </Wrapper>
  );
};

ErrorMessage.propTypes = {
  episodeAvailability: oneOf(Object.values(EPISODE_STATUS)).isRequired,
  skin: oneOf(['audio', 'video']),
};

ErrorMessage.defaultProps = {
  skin: 'video',
};

export default ErrorMessage;
