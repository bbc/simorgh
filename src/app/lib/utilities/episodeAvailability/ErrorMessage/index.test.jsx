/* eslint-disable react/prop-types */
import React from 'react';
import { render } from '@testing-library/react';
import { ServiceContextProvider } from '#contexts/ServiceContext';

import { EPISODE_STATUS } from '..';
import ErrorMessage from '.';

const ErrorMessageWithContext = ({ episodeAvailability, skin }) => (
  <ServiceContextProvider service="pidgin">
    <ErrorMessage skin={skin} episodeAvailability={episodeAvailability} />
  </ServiceContextProvider>
);

describe(`episodeAvailability - Error Message`, () => {
  it('should correctly render for future episodes with video skin', () => {
    const { getByText, container } = render(
      <ErrorMessageWithContext
        episodeAvailability={EPISODE_STATUS.EPISODE_IS_NOT_YET_AVAILABLE}
        skin="video"
      />,
    );

    expect(
      getByText('De thing wey de here never ready for you to play.'),
    ).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('should correctly render for expired episodes with audio skin', () => {
    const { getByText, container } = render(
      <ErrorMessageWithContext
        episodeAvailability={EPISODE_STATUS.EPISODE_IS_EXPIRED}
        skin="audio"
      />,
    );

    expect(getByText('Dis thing no dey again')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
