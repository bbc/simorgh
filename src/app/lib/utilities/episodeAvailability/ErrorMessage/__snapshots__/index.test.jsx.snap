// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`episodeAvailability - Error Message should correctly render for expired episodes with audio skin 1`] = `
.emotion-0 {
  position: relative;
  min-height: 165px;
  margin-bottom: 2rem;
}

.emotion-2 {
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-style: normal;
  font-weight: 400;
  font-size: 0.9375rem;
  line-height: 1.125rem;
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  left: 0;
  border: 0.0625rem solid transparent;
  color: #FFFFFF;
  background-color: rgba(34, 34, 34, 0.75);
}

@media (min-width: 20rem) and (max-width: 37.4375rem) {
  .emotion-2 {
    font-size: 0.9375rem;
    line-height: 1.125rem;
  }
}

@media (min-width: 37.5rem) {
  .emotion-2 {
    font-size: 0.875rem;
    line-height: 1.125rem;
  }
}

@media screen and (forced-colors: active) {
  .emotion-2 {
    background-color: transparent;
  }
}

.emotion-3 {
  display: block;
  font-weight: normal;
  bottom: 0;
  position: absolute;
  padding: 0.5rem;
}

@media (min-width: 25rem) {
  .emotion-3 {
    padding: 1rem;
  }
}

@media screen and (forced-colors: active) {
  .emotion-3 {
    background-color: window;
  }
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <div>
      <div
        class="emotion-2"
      >
        <strong
          class="emotion-3"
        >
          Dis thing no dey again
        </strong>
      </div>
    </div>
  </div>
</div>
`;

exports[`episodeAvailability - Error Message should correctly render for future episodes with video skin 1`] = `
.emotion-0 {
  margin: 2.5rem 0 1.5rem;
  padding-top: 56.25%;
  position: relative;
  overflow: hidden;
}

@media (max-width: 62.9375rem) {
  .emotion-0 {
    margin-top: 1rem;
  }
}

@media (max-width: 37.4375rem) {
  .emotion-0 {
    margin: 1rem -1rem 0;
  }
}

@media (max-width: 24.9375rem) {
  .emotion-0 {
    margin: 0.5rem -0.5rem 0;
  }
}

.emotion-2 {
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-style: normal;
  font-weight: 400;
  font-size: 0.9375rem;
  line-height: 1.125rem;
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  left: 0;
  border: 0.0625rem solid transparent;
  color: #FFFFFF;
  background-color: rgba(34, 34, 34, 0.75);
}

@media (min-width: 20rem) and (max-width: 37.4375rem) {
  .emotion-2 {
    font-size: 0.9375rem;
    line-height: 1.125rem;
  }
}

@media (min-width: 37.5rem) {
  .emotion-2 {
    font-size: 0.875rem;
    line-height: 1.125rem;
  }
}

@media screen and (forced-colors: active) {
  .emotion-2 {
    background-color: transparent;
  }
}

.emotion-3 {
  display: block;
  font-weight: normal;
  bottom: 0;
  position: absolute;
  padding: 0.5rem;
}

@media (min-width: 25rem) {
  .emotion-3 {
    padding: 1rem;
  }
}

@media screen and (forced-colors: active) {
  .emotion-3 {
    background-color: window;
  }
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <div>
      <div
        class="emotion-2"
      >
        <strong
          class="emotion-3"
        >
          De thing wey de here never ready for you to play.
        </strong>
      </div>
    </div>
  </div>
</div>
`;
